### PR TITLE
rows: Support all optional interfaces for driver.Rows

### PR DESCRIPTION
**Pull request by tcolgate**
This uses some code generation to wrap an incoming rows in a type
that supports the exact set of interfaces that are supported by
the original type.

### CODE REVIEW

**Comment by tcolgate** [comment]
This adapts a technique used in prometheus client_golang but uses code generation due to the number of combinations of optional types permitted by driver.Rows.
Unfortunately this approach does seem to be needed to at least the mysql,pgx and sqlite3 drivers which all support different combinations of these optional interfaces.

**Comment by tcolgate** [comment]
This poses a further interesting challenge. If the user wraps the Rows type (we are doing this with the goal of instrumenting row reads and open / close duration), The type we get back is itself wrapped and suffers from the same problem as this PR solves for the sqlmw rows type.
  One option would be to allow user wrapped Rows to implement an `func RowsUnwrap() driver.Rows`, that returns the original driver.Rows. We could check for that, call it if present and use the returned rows to probe the supported methods for wrapping.
    Users would then implement all the possible methods on their wrapper (passing calls through if present), but sqlmw would only expose a rows that supports the capabilities of the driver. 
  It might be better to leave that to a separate PR, to discuss API.

File: tools/rows_picker_gen.go
Comment (inconshreveable) [inline tools/rows_picker_gen.go:60]: remove?

File: rows_picker.go
Comment (inconshreveable) [inline rows_picker.go:1]: this file should have an appropriate generation comment at the top of it

File: rows_picker.go
Comment (inconshreveable) [inline rows_picker.go:980]: this can be defined in a non-generated file. it also needs to be documented for users of the library

**Review by inconshreveable — COMMENTED** [review]
Thank you for this contribution! Can we add a few tests that exercise the new functionality and prove its correctness?

File: tools/rows_picker_gen.go
Comment (tcolgate) [reply tools/rows_picker_gen.go:60]: > Thank you for this contribution! Can we add a few tests that exercise the new functionality and prove its correctness?

yes, I think so. The Go mysql, pgx and sqlite3 drivers expose different combiantions of these, so I was thinking of having mock drivers that match what they use. 
  It'll take me a while to get the tests sorted, didn't want to start the effort until I was sure you'd be interested in the merge.

File: rows_picker.go
Comment (inconshreveable) [reply rows_picker.go:1]: sure, but the comment should indiciate that the file was generated, the command that generated it and not to edit the file manually

**Comment by tcolgate** [comment]
Can you have a read of the RowsUnwrapper comment, I think it could be improved.
Assuming you are happy with these changes I'll start working on the tests tomorrow.

File: rows.go
Comment (fho) [inline rows.go:29]:
```suggestion
	Unwrap() driver.Rows
```

Comment (tcolgate) [reply rows.go:29]: Since this explicitly unwraps the rows type. and similar techniques could be applied elsewhere (like Tx) , being explicit felt like it made more sense here.

Comment (inconshreveable) [reply rows.go:29]: @fho's suggestion here feels right to me

re the comment: it could use some work, but it's directionally right.
- let's explain *why* you would wrap a Rows object
- "sqlmw will then pass a type that matches the optional interface set of the driver.Rows implementation of the original driver." this is hard to parse and could use clarity, i don't understand what it's trying to say
- maybe give an example method for the optional methods when it comes to wrapping rows objects

**Comment by tcolgate** [comment]
Updated the doc string. Better I think, but if you have specific suggestions...

**Comment by fho** [comment]
Thanks a lot for the PR. That is a great improvement.

1. According to the README Golang 1.9 is supported (which actually isn't in master: `./interceptor.go:17:36: undefined: driver.Connector`).
The rows_picker_gen.go from this PR is also not compatible with Go 1.9:
    ```go
    tools/rows_picker_gen.go:77:10: invalid operation: 1 << i (shift count type int, must be unsigned integer)
    tools/rows_picker_gen.go:85:12: invalid operation: 1 << len(intfs) (shift count type int, must be unsigned integer)
    ```
    This is trivial to fix.
Imho support for ancient go versions should be dropped.
Compatibility with the last 2-4 versions should be sufficient.
This is another topic though. :-)

2. I like the `Unwrap` approach.
If I understand the idea right, we could introduce in follow-up PRs similiar `TxUnwrapper`, `StmtUnwrapper`, `ConnUnwrapper`, `ResultUnwrapper` interfaces to make it work for all interceptor consistently? 
To make the same approach work for Stmts, the type of the stmt argument in the interceptor `Stmt*` methods still would need to be changed to not run into the issue that is described in https://github.com/ngrok/sqlmw/pull/16, right?

**Comment by tcolgate** [comment]
@inconshreveable @fho 
I've pushed a first stab at the tests. They aren't great but...
- Extended the exists tests to stress RowsNext more
- Added a direct test of wrapRows, testing internal functions isn't generally great, but this is the easiest way to ensure the interfaces match up.
- Added a more complete test of a pgx-like fake driver that ensure the rows description stuff gets called.

I wont be able to act on feedback until Wednesday, but hopefully we are close.

**Review by inconshreveable — APPROVED** [review]
LGTM. will merge pending @fho's approval

File: conn.go
Comment (fho) [inline conn.go:97]: If I understand it right, the nil context that is passed here is then forwarded to the intercepted rows methods?
If this is the case this will cause nil ptr dereference panics for users.
Users of the lib can assume that ctx is never nil because the stdlib context document says it must never be nil. (https://pkg.go.dev/context)
```suggestion
		return wrapRows(context.Background(), c.intr, rows), nil //nolint
```

File: rows.go
Comment (fho) [inline rows.go:13]: optional:
To make it consistent, we should add similiar assignments for all newly added wrappedRows types, alternatively we could also remove the last remaining `_ driver.Rows = wrappedRows{}` assignment.

File: tools/rows_picker_gen.go
Comment (fho) [inline tools/rows_picker_gen.go:26]: The return value of Close() should be checked, if it is not nil the program should exit with an error.
In the unlikely case that Close() failed the content might have not be written to disk.

File: rows_test.go
Comment (fho) [inline rows_test.go:327]: optional:

```suggestion
				t.Fatalf("inconsistent support for driver.RowsNextResultSet")
```
That typo occurs multiple times. :-)

File: rows_test.go
Comment (fho) [inline rows_test.go:101]: optional:

If `go test` is called with a `-count` parameter with a value >1 the testcases will fail because a driver with the same name is registered multiple times.
The test name is not unique enough, adding a microsec timestamp or a uuid could fix it.

The issue is not introduced with this PR and exists also in other testcases.
Therefore I don't see it as strictly necessary to be changed for the new testcases in this PR.

File: rows_test.go
Comment (fho) [inline rows_test.go:129]:
```suggestion
			t.Fatal(err)
```

File: rows_test.go
Comment (fho) [inline rows_test.go:258]:
```suggestion
			t.Fatal(err)
```

File: rows_test.go
Comment (fho) [inline rows_test.go:229]:
```suggestion
	// There's no real way these can be called, but we'll test in case the
```

File: rows_test.go
Comment (tcolgate) [reply rows_test.go:101]: One approach I've used for this kind of thing before is to have an atomic somewhere that you increment when generating a driver name.

Comment (fho) [reply rows_test.go:101]: that approach is much better then my suggestion :-)

**Review by fho — APPROVED** [review]

**Comment by inconshreveable** [comment]
ready to merge after tests pass

**Comment by tcolgate** [comment]
I will finish the review comments on Tuesday

**Comment by tcolgate** [comment]
Hopefully all good to go now?

**Comment by inconshreveable** [comment]
ty @tcolgate !

**Pull request merged**
